src/Translator/Extractor/AdminExtractor.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Translator\Extractor;
+
+use Psr\Log\LoggerInterface;
+use Sonata\AdminBundle\Admin\BreadcrumbsBuilderInterface;
+use Sonata\AdminBundle\Admin\Pool;
+use Sonata\AdminBundle\Translator\LabelTranslatorStrategyInterface;
+use Symfony\Component\Translation\Extractor\ExtractorInterface;
+use Symfony\Component\Translation\MessageCatalogue;
+
+/**
+ * @internal
+ */
+final class AdminExtractor implements ExtractorInterface, LabelTranslatorStrategyInterface
+{
+    /**
+     * @var string
+     */
+    private $prefix;
+
+    /**
+     * @var MessageCatalogue|null
+     */
+    private $catalogue;
+
+    /**
+     * @var LoggerInterface|null
+     */
+    private $logger;
+
+    /**
+     * @var Pool
+     */
+    private $adminPool;
+
+    /**
+     * @var LabelTranslatorStrategyInterface|null
+     */
+    private $labelStrategy;
+
+    /**
+     * @var string|null
+     */
+    private $domain;
+
+    /**
+     * @var BreadcrumbsBuilderInterface
+     */
+    private $breadcrumbsBuilder;
+
+    public function __construct(Pool $adminPool, BreadcrumbsBuilderInterface $breadcrumbsBuilder, LoggerInterface $logger = null)
+    {
+        $this->logger = $logger;
+        $this->adminPool = $adminPool;
+        $this->breadcrumbsBuilder = $breadcrumbsBuilder;
+    }
+
+    public function extract($resource, MessageCatalogue $catalogue)
+    {
+        $this->catalogue = $catalogue;
+
+        foreach ($this->adminPool->getAdminGroups() as $name => $group) {
+            $catalogue->set($name, $this->prefix.$name, $group['label_catalogue']);
+        }
+
+        foreach ($this->adminPool->getAdminServiceIds() as $id) {
+            $admin = $this->adminPool->getInstance($id);
+
+            $this->labelStrategy = $admin->getLabelTranslatorStrategy();
+            $this->domain = $admin->getTranslationDomain();
+
+            $label = $admin->getLabel();
+            if (!empty($label)) {
+                $catalogue->set($label, $this->prefix.$label, $admin->getTranslationDomain());
+            }
+
+            $admin->setLabelTranslatorStrategy($this);
+
+            // call the different public method
+            $methods = [
+                'getShow',
+                'getDatagrid',
+                'getList',
+                'getForm',
+            ];
+
+            $actions = [
+                'list',
+                'edit',
+                'create',
+                'update',
+                'batch',
+                'delete',
+            ];
+
+            if ($this->logger) {
```

Comment:
If the injected logger is `null`, please set `$this->logger` to a `NullLogger` instance. It will allow you to remove such if statements

Reply:
I'll take a look, but maybe I can remove the logger completely